src/ray/gcs/gcs_server/gcs_node_manager.cc

```diff
@@ -224,11 +224,62 @@ absl::optional<std::shared_ptr<rpc::GcsNodeInfo>> GcsNodeManager::GetAliveNode(
   return iter->second;
 }
 
-void GcsNodeManager::SetDeathInfo(const NodeID &node_id, rpc::NodeDeathInfo death_info) {
+void GcsNodeManager::SetDeathInfo(const NodeID &node_id,
+                                  shared_ptr<rpc::NodeDeathInfo> death_info) {
   auto maybe_node = GetAliveNode(node_id);
-  RAY_CHECK(maybe_node.has_value());
+  if (!maybe_node.has_value()) {
+    RAY_LOG(INFO) << "Skip setting death info for node " << node_id
+                  << " which is already removed.";
+    return;
+  }
+
+  RAY_CHECK(death_info) << "death_info should not be nullptr";
   auto node = std::move(maybe_node.value());
-  node->mutable_death_info()->CopyFrom(death_info);
+  auto node_death_info = node->mutable_death_info();
+  node_death_info->CopyFrom(*death_info);
+  RAY_LOG(INFO) << "Setting death info for node " << node_id << ": Reason = "
+                << rpc::NodeDeathInfo_Reason_Name(node_death_info->reason())
+                << ", message = " << node_death_info->reason_message();
+}
+
+shared_ptr<rpc::NodeDeathInfo> GcsNodeManager::InferDeathInfo(const NodeID &node_id,
+                                                              bool &caused_by_draining) {
+  auto maybe_node = GetAliveNode(node_id);
+  RAY_CHECK(maybe_node.has_value())
+      << "InferDeathInfo() should be called before node is removed";
+  auto node = maybe_node.value();
+  auto iter = draining_nodes_.find(node_id);
+  auto death_info = std::make_shared<rpc::NodeDeathInfo>();
+  if (iter == draining_nodes_.end()) {
+    caused_by_draining = false;
+  } else if (iter->second->deadline_timestamp_ms() == 0) {
+    // If there is no draining deadline, there should not be force termination
+    caused_by_draining = false;
+  } else {
+    caused_by_draining = current_sys_time_ms() > iter->second->deadline_timestamp_ms();
+  }
+
+  RAY_LOG(INFO) << "For the death of node " << node_id
+                << ", inferred caused_by_draining = " << caused_by_draining;
+
+  if (caused_by_draining) {
+    std::shared_ptr<rpc::autoscaler::DrainNodeRequest> drain_request = iter->second;
```

Comment:
Actually after your refactor, seems this shard_ptr doesn't provide any benefit now?

Reply:
Yeah, fixed.